### PR TITLE
add uv_queue_work_idle_threads to return how many idle threads are pr…

### PR DESCRIPTION
**Pull request by bsergean**
…esent in the thread pool

### CODE REVIEW

File: include/uv.h [hunk 1]
```diff
@@ -1100,6 +1100,8 @@ UV_EXTERN int uv_queue_work(uv_loop_t* loop,
 
 UV_EXTERN int uv_cancel(uv_req_t* req);
 
+UV_EXTERN unsigned int uv_queue_work_idle_threads();
```
Comment (bnoordhuis) [inline include/uv.h:1103]:
```suggestion
UV_EXTERN unsigned int uv_queue_work_idle_threads(void);
```

File: src/threadpool.c [hunk 2]
```diff
@@ -160,6 +160,11 @@ static void post(QUEUE* q, enum uv__work_kind kind) {
 }
 
 
+unsigned int uv_queue_work_idle_threads(void) {
+    return idle_threads;
```
Comment (bnoordhuis) [inline src/threadpool.c:164]:
```suggestion
  return idle_threads;
```

**Review by bnoordhuis — COMMENTED** [review]
Thanks for the PR but I don't really understand when or why you'd need that info.

**Comment by bsergean** [comment]
We use this as a metric, to look at the thread-pool utilization ; lots of idle threads means the thread-pool isn't very active, and vice versa.

We periodically call this function, store the result in statsd etc..., then chart it in grafana (https://grafana.com/)

**Comment by bsergean** [comment]
To be more specific, we had a bug where we were bombarding the thread-pool with jobs, but we didn't know and use that new function to figure it out.

We understand it's quite peculiar and not of interest to most people, but a few might find it useful, and the flipside is that it's a pretty simple change so it should not break anything.

**Comment by bnoordhuis** [comment]
Libuv takes backwards compatibility seriously so once we add this, we can't remove it again.

It exposes what is essentially an implementation detail (meaning it might constrain future redesign) so I'm not eager to merge it but I'll leave the PR open to gauge interest from others.

**Comment by bsergean** [comment]
Understood. Do you see any ways with the existing api, or with a newer one, which does not expose implementation details as much, to expose thread pool utilization ?

I see uv_metrics_idle_time which looks interesting: http://docs.libuv.org/en/v1.x/metrics.html

Also we have another change in our fork to name the threads from the threadpool (with api like this one https://www.ibm.com/docs/en/i/7.2?topic=ssw_ibm_i_72/apis/pthread_attr_setname_np.htm). Can that be used to identify the threads, and call a fancy function to get utilization for those threads ? (I didn't find such functions, the closest is uv_getrusage but it's for the whole process).

**Comment by bsergean** [comment]
One quick reply to have this PR survive a little longer but if there isn’t much interest it’s ok to close it.

Given the simplicity of the change, we could have a little patch that we apply before we build, or similar.

**Comment by bsergean** [comment]
Ok let's give up on this.

**Pull request closed without merging**